tests/baselines/reference/parserForInStatement8.errors.txt

```diff
@@ -0,0 +1,20 @@
+parserForInStatement8.ts(3,10): error TS2461: Type 'string' is not an array type.
+parserForInStatement8.ts(3,10): error TS2491: The left-hand side of a 'for...in' statement cannot be a destructuring pattern.
+parserForInStatement8.ts(4,10): error TS2491: The left-hand side of a 'for...in' statement cannot be a destructuring pattern.

```

Comment:
Those errors are not correct but that's a separate issue

Reply:
How challenging is that to fix?

Reply:
I'm looking into this but it's the first time that I even opened the files related to downleveling syntax here so it might take a moment (if I don't lose interest in this, that is). 

1. the error has to be likely removed completely (I removed it from one location but there is another one and I didn't even look into it yet)
2. [this `undefined` shouldn't be emitted](https://github.com/microsoft/TypeScript/blob/a3773ec590c4f0308d546f0e65818cd0d12402f3/src/compiler/transformers/destructuring.ts#L354-L355) (at least in this case)
3. and the declaration list can't be "replaced" in place, since downleveling binding patterns requires creating temp bindings and for-in only supports a single binding (`Uncaught SyntaxError: Invalid left-hand side in for-in loop: Must have a single binding.`)
4. to fix 2 and 3 it's likely the easiest to move the binding "expansion" to the loop's body, this is what Babel does:
```ts
for (var _ref3 in {
  '': 0
}) {
  var _ref2 = _toArray(_ref3);
  var _ref2$ = _ref2[0];
  var x = _ref2$ === void 0 ? 'a' in {} : _ref2$;
  var kaka = _ref2[1];
  var rest = _ref2.slice(2);
  console.log(x);
}
```

Reply:
It’s legal to use destructuring in `for..in`?  What would that even *mean* since the value is always a string (or symbol)?

Reply:
Strings are iterable so u could read its characters through an array pattern, or some prototype methods/properties through an object pattern.

Is it useful? Not rly. Is it legal? Apparently yes :P

Reply:
> Is it useful? Not rly. Is it legal? Apparently yes :P

Hmm, in that case it might be worth keeping in mind https://stackoverflow.com/questions/41750390/what-does-all-legal-javascript-is-legal-typescript-mean

Reply:
Eh, that's more for type errors. Generally speaking, our parser needs to parse all valid JS because tons of people use it to parse JS.

The only exception in recent memory is us dropping non-strict parsing of certain escapes in certain literals (as that came with a huge perf advantage among many bugfixes).

Reply:
Of course, if there's no issue for this already, that's probably pretty telling that we don't need to try and fix it, if this code was already not parsing before.

Reply:
After a second thought, the array pattern is actually useful and it's supported by TypeScript in other contexts. The object pattern is not really useful, IMHO, but it works in TS. It would be weird to support one type of a pattern here without the other one and I think it's worth fixing to be more correct and consistent. [TS playground](https://www.typescriptlang.org/play?target=99#code/PTAEAEBcEMCcHMCmkBcpEGcB2iAekAoAYwHssNJQBtAMwEtYKAaUAOnZjoBsBdUAXlABySJkhCC9RoRCgAegH4CnLgVmKCxMhVABvUF0RZ4kABZpDxswEZQAXwHDRFCZZOnrasIqA)

I have a fix for this locally, just need to recheck those changes with the checker since I only worried about the es2015 transform so far.

Reply:
FWIW I think people sometimes write `{ 0: first, 1: second }` instead of `[ first, second ]` to avoid the performance hit of calling the iterator.

Reply:
Right, it might differ between engines but IIRC the iterator overhead was largely patched up in v8 when arrays are involved. There was a nice blog post about this optimization.

Reply:
Just close the loop on this one, if you don't think you'll get around to fixing this in this PR, can you add a note to the test case that says "hey, this is wrong" and file an issue, or similar?

Reply:
Er, actually, is #54853 that PR?

Reply:
Yes, it is.

<img src="https://media.giphy.com/media/3og0IxfCJH5VTEJQC4/giphy.gif" alt="Gif saying 'You taught me that, Saul. No loose ends'"/>

Reply:
Alright, I guess we can just leave the PR as-is?

Reply:
Yes, i think so.